solidity/test/bridge/VendingMachine.Upgrade.test.ts

```diff
@@ -0,0 +1,222 @@
+    ;({ tbtcVault, tbtc, vendingMachine, bank, bridge, relay } =
+      await waffle.loadFixture(bridgeFixture))
+
+    // Set the deposit dust threshold to 0.0001 BTC, i.e. 100x smaller than
+    // the initial value in the Bridge in order to save test Bitcoins.
+    await bridge.setDepositDustThreshold(10000)
+
+    await bridge.connect(governance).setVaultStatus(tbtcVault.address, true)
+
+    tbtcV1 = await helpers.contracts.getContract("TBTCToken")
+    // Two accounts with 10 TBTC v1 each wrap their holdings to TBTC v2.
+    const initialTbtcBalance = to1e18(10)
+    await tbtcV1.connect(deployer).mint(account1.address, initialTbtcBalance)
+    await tbtcV1.connect(deployer).mint(account2.address, initialTbtcBalance)
+    await tbtcV1
+      .connect(account1)
+      .approveAndCall(vendingMachine.address, initialTbtcBalance, [])
+    await tbtcV1
+      .connect(account2)
+      .approveAndCall(vendingMachine.address, initialTbtcBalance, [])
+
+    // Deployment scripts deploy both `VendingMachine` and `TBTCVault` but they
+    // do not transfer the ownership of `TBTC` token to `TBTCVault`.
+    // We need to do it manually in tests covering `TBTCVault` behavior.
+    // Also, please note that `03_transfer_roles.ts` assigning `VendingMachine`
+    // upgrade initiator role to Keep Technical Wallet is skipped for Hardhat
+    // env deployment. That's why the upgrade initiator and `VendingMachine`
+    // owner is the deployer.
+    await vendingMachine
+      .connect(deployer)
+      .initiateVendingMachineUpgrade(tbtcVault.address)
+    await increaseTime(await vendingMachine.GOVERNANCE_DELAY())
+    await vendingMachine.connect(deployer).finalizeVendingMachineUpgrade()
+  })
+
+  describe("upgrade process", () => {
+    // Two accounts with 10 TBTC v1 each wrap their holdings to TBTC v2.
+    // See the main `before`.
+    const totalTbtcV1Balance = to1e18(20)
+
+    describe("step#1 - TBTC v1 transfer", () => {
+      it("should transfer all TBTC v1 to TBTCVault", async () => {
+        expect(await tbtcV1.balanceOf(vendingMachine.address)).to.equal(
+          to1e18(0)
+        )
+        expect(await tbtcV1.balanceOf(tbtcVault.address)).to.equal(
+          totalTbtcV1Balance
+        )
+      })
+    })
+
+    describe("step#2 - TBTC v1 withdrawal", () => {
+      it("should let the governance withdraw TBTC v1 from TBTCVault", async () => {
+        await tbtcVault
+          .connect(governance)
+          .recoverERC20(tbtcV1.address, governance.address, totalTbtcV1Balance)
+        expect(await tbtcV1.balanceOf(tbtcVault.address)).to.equal(0)
+        expect(await tbtcV1.balanceOf(governance.address)).to.equal(
+          totalTbtcV1Balance
+        )
+      })
+    })
+
+    describe("step#3 - BTC donation", () => {
+      before(async () => {
+        const data = SingleP2SHDeposit
+        const { fundingTx, reveal } = data.deposits[0] // it's a single deposit
+        reveal.vault = tbtcVault.address
```

Comment:
If we do `reveal.vault = tbtcVault.address` that way, we modify the original `SingleP2SHDeposit` object which may give surprises during bulk test execution. In order to prevent it, let's do a copy of `SingleP2SHDeposit` first:
```
const data: DepositSweepTestData = JSON.parse(JSON.stringify(SingleP2SHDeposit)) as Deposit
```

Reply:
7b32074783820a4ef0d9e286ab2d09ad896f57f1